### PR TITLE
Fb infer

### LINKED ISSUE #4189: Facebook Infer analyzer support
Add support for facebook infer analyzer execution support for C/C++.
https://github.com/facebook/infer

CodeChecker should be able to 
* List checkers: CodeChecker checkers
* Checkers should have "infer" prefix: infer.biabduction_memory_leak
* Execute analysis: `CodeChecker analyze --anlayzers infer ./compile_commands.json`
* Enable/disble checkers for the analysis:` CodeChecker analyze --analyzers infer -e infer.biabduction_memory_leak`
* Enable storing analysis results `CodeChecker store ./report-dir --url https://...`
* Document how the analyzer is supported in 
** https://github.com/Ericsson/codechecker/blob/master/docs/analyzer/checker_and_analyzer_configuration.md
** https://github.com/Ericsson/codechecker/blob/fa1d14f624414056b6ad2d04454b88060d18c2f4/docs/analyzer/user_guide.md
** https://github.com/Ericsson/codechecker/tree/fa1d14f624414056b6ad2d04454b88060d18c2f4/docs#readme

### PR DESCRIPTION
**Pull request by stt08**
Adding facebook infer analyzer support
FIXES #4189

### CODE REVIEW

**Comment by vodorok** [comment]
I've not tried it, but this integration should support the new 1.2.0 release of infer.
https://github.com/facebook/infer/releases/tag/v1.2.0

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 1]
```diff
@@ -300,3 +300,29 @@ CodeChecker check -l ./compile_commands.json \
   -d gcc-double-free \ # disable gcc-double-free
   -o ./reports
 ```
+
+# Configuring the FB-Infer Analyzer
+
+As of CodeChecker 6.23, Codechecker can now execute the Facebook Infer Analyzer.
+The minimum version of GCC we support is 1.0.0.
```
Comment (dkrupp) [inline docs/analyzer/checker_and_analyzer_configuration.md:307]: minimum version of facebook infer! is 1.1.0??

**Review by dkrupp — CHANGES_REQUESTED** [review]
Please check why the analysis does not work on postgresql
Please fix the description field of the infer checkers
There are other minor copy paste errors which you need to check

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 2]
```diff
@@ -300,3 +300,29 @@ CodeChecker check -l ./compile_commands.json \
...
+## Limitations
+
+Currently only static analysis can be executed. Meaning that it analyzes each
+file separately and not the whole project as one.
+
+## Example invocation
+
+``` shell
+CodeChecker check -l ./compile_commands.json \
+  --analyzers infer \ # Run GCC analyzer only
```
[the remark this answers is not in the document]
Comment (stt08) [reply docs/analyzer/checker_and_analyzer_configuration.md]: fixed

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 3]
```diff
@@ -300,3 +300,29 @@ CodeChecker check -l ./compile_commands.json \
...
+  -e infer \ # enable all checkers starting with "gcc"
```
[the remark this answers is not in the document]
Comment (stt08) [reply docs/analyzer/checker_and_analyzer_configuration.md]: fixed

File: docs/examples/compile_commands.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
+[
+	{
+		"directory": "/home/t/codechecker/docs/examples",
+		"command": "/usr/bin/cc -MMD -MP -g -std=c99 -Wall -I /home/t/codechecker/docs/examples/incl -c /home/t/codechecker/docs/examples/src/divide.c -o /home/t/codechecker/docs/examples/objs/divide.c.o",
```
[the remark this answers is not in the document]
Comment (stt08) [reply docs/examples/compile_commands.json]: fixed: file was added accidentally, it is now removed

File: analyzer/codechecker_analyzer/analyzers/infer/analyzer.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+        # analyzer_cmd.append(self.source_file)
+
+        LOG.debug_analyzer("Running analysis command "
+                           f"'{shlex.join(analyzer_cmd)}'")
+
+        return analyzer_cmd
+
+    @classmethod
+    def get_analyzer_checkers(self):
```
[the remark this answers is not in the document]
Comment (stt08) [reply analyzer/codechecker_analyzer/analyzers/infer/analyzer.py:95]: fixed: description have been parsed from the website into a json file and is being replaced by id

File: docs/analyzer/checker_and_analyzer_configuration.md (hunk 1 above)
Comment (stt08) [reply docs/analyzer/checker_and_analyzer_configuration.md]: fixed

File: analyzer/codechecker_analyzer/analyzers/infer/descriptions.json [hunk 6]
```diff
@@ -0,0 +1,200 @@
+{
```
Comment (dkrupp) [inline analyzer/codechecker_analyzer/analyzers/infer/descriptions.json:1]: Please include the fb infer original license file
https://github.com/facebook/infer/blob/main/LICENSE
with the additional description that it is valid for the descriptions.json file

**Review by dkrupp — CHANGES_REQUESTED** [review]
please fix the postgresql and the xerces analysis and add the fb infer license file.

**Comment by dkrupp** [comment]
I retried this PR with infer 1.1.0.

I could not list the checkers
```
CodeChecker checkers --analyzers infer
[ERROR 2024-08-26 17:08] - None
[ERROR 2024-08-26 17:08] - None
[INFO 2024-08-26 17:08] - No checkers with the given label found.
```

and I could not execute the analysis either, I guess because it could not list checker.

```
workspace/test-projects/xerces-c]$ CodeChecker analyze ./compile_commands.json --analyzers=infer -e infer -o ./reports-infer
[ERROR 2024-08-26 17:09] - None
[INFO 2024-08-26 17:09] - No checkers enabled for infer
[ERROR 2024-08-26 17:09] - No checker(s) with these names was found:
infer
```

I think the problem is that the timezone setting is incorrect for infer.
```
nfer
Uncaught Internal Error: ("unknown zone" (zone :Europe/Budapest))
Error backtrace:
Raised at Base__Error.raise in file "src/error.ml" (inlined), line 8, characters 14-30
Called from Base__Error.raise_s in file "src/error.ml", line 9, characters 19-40
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from IBase__ResultsDir.RunState.run_time_string in file "src/base/ResultsDir.ml", line 14, characters 24-53

Run the command again with `--keep-going` to try and ignore this error.
(CodeChecker venv-dev) ednikru@seliiuvd02700[14:47][build/CodeChecker/bin]$ echo $?
2
```

Could you please print the stderror from the analyzer tool in as CodeChecker error log when invocation of the tools fails?
This way the user could debug a failing analyzer binary.

Or course when I set `export TIMEZONE=UTC` in my shell, infer was working correctly through CodeChecker. So this is about error handling.

**Comment by dkrupp** [comment]
Infer 1.1.0 drops the following error if the timezone is not set to UTC:

```
infer
Uncaught Internal Error: ("unknown zone" (zone :Europe/Budapest))
Error backtrace:
Raised at Base__Error.raise in file "src/error.ml" (inlined), line 8, characters 14-30
Called from Base__Error.raise_s in file "src/error.ml", line 9, characters 19-40
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from IBase__ResultsDir.RunState.run_time_string in file "src/base/ResultsDir.ml", line 14, characters 24-53

```

This also means that the 
`CodeChecker checkers --analyzers infer`
lists nothing and the infer analyzer is unusable from under codechecker.

Please add the TZ=UTC environment variable where you invoke infer as a workaround.

```
TZ=UTC infer
Nothing to compile. Have you run `infer capture`? Try cleaning the build first.
There was nothing to analyze.

  No issues found  
```
like this, it works.

File: analyzer/codechecker_analyzer/analyzers/infer/analyzer.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        """
+        Return the analyzer version.
+        """
+        # No need to LOG here, we will emit a warning later anyway.
+        if not cls.analyzer_binary():
+            return None
+        version = [cls.analyzer_binary(), '--version']
+        try:
+            output = subprocess.check_output(version,
+                                             env={'TZ': 'UTC'},
```
Comment (dkrupp) [inline analyzer/codechecker_analyzer/analyzers/infer/analyzer.py:181]: Pelase initialize the environment like this https://github.com/Ericsson/codechecker/blob/321492ee2d2e25a0e2abc5910597678af535ba3f/analyzer/codechecker_analyzer/analyzers/cppcheck/analyzer.py#L90

and only extend the environment with TZ:UTC, not overwrite it.

**Review by dkrupp — APPROVED** [review]

**Pull request merged**